### PR TITLE
add custom memory access calls

### PR DESCRIPTION
**Pull request by philvukovic**
* change `MIPSImage::custom_memory_read` to return a `std::optional<mem_word>`
* add `MIPSImage::custom_memory_read` calls to `read_mem` functions
* add `MIPSImage::custom_memory_write` calls to `set_mem` functions

### CODE REVIEW

File: spim/CPU/mem.cpp [hunk 1]
```diff
@@ -388,6 +392,10 @@ read_mem_byte(MIPSImage &img, mem_addr addr)
 reg_word
 read_mem_half(MIPSImage &img, mem_addr addr)
 {
+  std::optional<reg_word> custom_read = img.custom_memory_read(addr);
+  if (custom_read.has_value())
+    return custom_read.value();
+
```
Comment (Nydauron) [inline spim/CPU/mem.cpp]: MMIO addresses should expect reading/writing a word. It definitely complicates the logic if we include this one function for all 3 different reads/writes.

What you can do is add a virtual `custom_memory_read_word()`, `custom_memory_read_half()`, `custom_memory_read_byte()`, ... same for the custom writes. It still allows for the functionality of doing custom writes for half words and bytes, but doesnt interfere with our logic of word-only MMIOs.

File: spim/CPU/mem.cpp [hunk 2]
```diff
@@ -372,6 +372,10 @@ read_mem_inst(MIPSImage &img, mem_addr addr)
 reg_word
 read_mem_byte(MIPSImage &img, mem_addr addr)
 {
+  std::optional<reg_word> custom_read = img.custom_memory_read(addr);
+  if (custom_read.has_value())
+    return custom_read.value();
+
```
Comment (Nydauron) [inline spim/CPU/mem.cpp]: Same as above comment

File: spim/CPU/mem.cpp [hunk 3]
```diff
@@ -434,6 +446,9 @@ void
 set_mem_byte(MIPSImage &img, mem_addr addr, reg_word value)
 {
   img.mem_image().data_modified = true;
+  if (img.custom_memory_write(addr, value))
+    return;
+
```
Comment (Nydauron) [inline spim/CPU/mem.cpp]: Same

File: spim/CPU/mem.cpp [hunk 4]
```diff
@@ -451,6 +466,9 @@ void
 set_mem_half(MIPSImage &img, mem_addr addr, reg_word value)
 {
   img.mem_image().data_modified = true;
+  if (img.custom_memory_write(addr, value))
+    return;
+
```
Comment (Nydauron) [inline spim/CPU/mem.cpp]: Same

**Review by Nydauron — CHANGES_REQUESTED** [review]

**Comment by philvukovic** [comment]
should be good now

**Review by Nydauron — APPROVED** [review]

**Pull request merged**